### ISSUE TITLE
createNewBackup() intermittently throws "checksum mismatch" with writes in parallel

### ISSUE THREAD

**Issue by ktnowak** [facebook/rocksdb#3266]
### Problem
We see following exception intermittently thrown from `BackupEngine.createNewBackup()` when there is high volume of batch-writes running in parallel.

```
org.rocksdb.RocksDBException: checksum mismatch
at org.rocksdb.BackupEngine.createNewBackup(Native Method)
at org.rocksdb.BackupEngine.createNewBackup(BackupEngine.java:81)
at org.rocksdb.BackupEngine.createNewBackup(BackupEngine.java:53)
at org.rocksdb.test.RocksDbChecksumError.testBackupCausesChecksumError(RocksDbChecksumError:44)
```

RocksDB 5.7.3 LOG excerpt
```
2017/12/3-22:27:36.148268 7fc9c410e700 [db/db_impl_write.cc:1127] [testCf] New memtable create with log file: #811. Immutable memtables 0.
(...)
2017/12/3-22:27:36.153696 7fc9c420e700 [db/filesnapshot.cc:34] File Deletions Disabled
2017/12/3-22:27:36.160151 7fc9c420e700 [WARN] [db/wal_manager.cc:425] /path/to/rocks/data/000811.log: dropping 32768 bytes; Corruption: checksum mismatch
2017/12/3-22:27:36.160161 7fc9c420e700 [WARN] [db/wal_manager.cc:425] /path/to/rocks/data/000811.log: dropping 2555358 bytes; Corruption: error in middle of record
2017/12/3-22:27:36.160199 7fc9c420e700 [WARN] [db/wal_manager.cc:425] /path/to/rocks/data/000811.log: dropping 32761 bytes; Corruption: missing start of fragmented record(1)
(...)
2017/12/3-22:27:36.169905 7fc9c420e700 [WARN] [db/wal_manager.cc:425] /path/to/rocks/data/000811.log: dropping 5988 bytes; Corruption: missing start of fragmented record(2)
2017/12/3-22:27:36.170401 7fc9c420e700 [db/filesnapshot.cc:57] File Deletions Enabled
(...)
2017/12/3-22:27:36.182490 7fc96c3f9700 [db/db_impl_compation_flush.cc:50] [JOB 1259] Syncing log #811
(...)
2017/12/3-22:27:36.204967 7fc96c3f9700 [db/db_impl_files.cc:397] [JOB 1259] Try to delete WAL files size 6034012, prev total WAL file size 12068024, number of live WAL files 2.
2017/12/3-22:27:36.205022 7fc96c3f9700 [db/wal_manager.cc:269] [JOB 1259] Move log file /path/to/rocks/data/000811.log to /path/to/rocks/data/archive/000811.log -- OK
```

After exception occurs, all subsequent attempts to create backup will consitently fail with the same error complaining about corrupted WAL log.
[](url)
It is also intermittently raised when trying to create checkpoint during similar circumstances (what is done by BackupEngine under the hood).

### Environment
RHEL 6.7 x86_64 4xCPU
RocksDB 5.7.3, 5.7.5, 5.8.7

### Steps to reproduce the behavior
Execute Junit tests on Linux env from the package attached:
`RocksDbChecksumError.testBackupCausesChecksumError()`
`RocksDbChecksumError.testChecpointCausesChecksumError()`

[rocksdb-checksumerror.zip](https://github.com/facebook/rocksdb/files/1554021/rocksdb-checksumerror.zip)

**Comment by maysamyabandeh** [comment]
Thanks for the report. Assuming that you do not call GetUpdatesSince, the printed error must be from this stack trace:
PurgeObsoleteFiles -> GetSortedWalsOfType -> ReadFirstRecord -> ReadFirstLine

**Comment by maysamyabandeh** [comment]
The candidate stack trace is odd since when the file deletions is disabled PurgeObsoleteFiles should not be invoked:
```
538 void DBImpl::DeleteObsoleteFiles() {
539   mutex_.AssertHeld();
540   JobContext job_context(next_job_id_.fetch_add(1));
541   FindObsoleteFiles(&job_context, true);
542
543   mutex_.Unlock();
544   if (job_context.HaveSomethingToDelete()) {
545     PurgeObsoleteFiles(job_context);
546   }
```
where
```142 void DBImpl::FindObsoleteFiles(JobContext* job_context, bool force,
143                                bool no_full_scan) {
144   mutex_.AssertHeld();
145
146   // if deletion is disabled, do nothing
147   if (disable_delete_obsolete_files_ > 0) {
148     return;
149   }
```

**Comment by maysamyabandeh** [comment]
I suspect this is a bug that is already fixed in 
```
commit 01542400a87ad130aed790bf895c029082745cbe
Author: Adrien Schildknecht <adrs@fb.com>
Date:   Thu Oct 5 18:00:38 2017 -0700

    Inform caller when rocksdb is stalling writes
```
Before HaveSomethingToDelete would return true even if delete obsolete files was disabled.
```
 22   inline bool HaveSomethingToDelete() const {
 23     return full_scan_candidate_files.size() || sst_delete_files.size() ||
 24            log_delete_files.size() || manifest_delete_files.size() ||
 25            new_superversion != nullptr || superversions_to_free.size() > 0 ||
 26            memtables_to_free.size() > 0 || logs_to_free.size() > 0;
 27   }
```
@ktnowak are you sure you got the same log with 5.8.3?

**Comment by ktnowak** [comment]
Log above was produced for RocksDB version 5.7.3.
However we see the problem for version 5.8.7 as well - please see excerpt below.

Were you able to reproduce the error after running java tests in packed into [rocksdb-checksumerror.zip](https://github.com/facebook/rocksdb/files/1554021/rocksdb-checksumerror.zip)? 
(you may tweak RockDB version in pom.xml)

RockDB 5.8.7 LOG
```
2017/12/15-18:45:50.844117 7f6e10541700 [db/db_impl_write.cc:1150] [testCf] New memtable create with log file: #7695. Immutable memtables 0.
(...)
2017/12/15-18:45:50.844565 7f6e1298a700 [db/filesnapshot.cc:34] File Deletions Disabled
(...)
2017/12/15-18:45:50.846550 7f6e11e88700 [db/flush_job.cc:264] [testCf] [JOB 15695] Flushing memtable with next log file: 7695
(...)
2017/12/15-18:45:50.865205 7f6e1298a700 [WARN] [db/wal_manager.cc:426] [WalManager]  /path/to/rocks/data/0007695.log: dropping 32768 bytes; Corruption: checksum mismatch
2017/12/15-18:45:50.865215 7f6e1298a700 [WARN] [db/wal_manager.cc:426] [WalManager]  /path/to/rocks/data/0007695.log: dropping 98283 bytes; Corruption: error in middle of record
2017/12/15-18:45:50.866340 7f6e1298a700 [WARN] [db/wal_manager.cc:426] [WalManager]  /path/to/rocks/data/0007695.log: dropping 32768 bytes; Corruption: checksum mismatch
2017/12/15-18:45:50.868026 7f6e1298a700 [db/db_filesnapshot.cc:57] File Deletions Enabled
(...)
2017/12/15-18:45:50.884242 7f6e11e88700 [db/db_impl_compaction_flush.cc:50] [testCf] [JOB 15699] Syncing log #7695
(...)
2017/12/15-18:45:50.906355 7f6e11e88700 [db/db_impl_files.cc:400] [JOB 15699] Try to delete WAL files size 6034012, prev total WAL files size 12068024, number of live WAL files 2.
2017/12/15-18:45:50.906399 7f6e11e88700 [db/wal_manager:270] Move log file /path/to/rocks/data//0007695.log to /path/to/rocks/data/archive/0007695.log -- OK
```

**Comment by ktnowak** [comment]
Update - after building 5.9.2 locally and running the test attached, it turned out that occurrence rate of the issue has dropped dramatically (we had tests running for 10+ minutes w/o raising the error whereas on 5.7.3 or 5.8.7 it already failed after ~ 20sec or 1min 30 sec respectively).  Test runs were interrupted by running out of disk space rather than error itself.

Do you have some rough estimate when rocksdbjni-5.9.2.jar will be available in maven central?


That said, as we actually saw infrequent checksum errors on 5.9.2 as well. Attaching log excerpt for your reference:
```
2017/12/19-18:53:42.023690 7f05d0765700 RocksDB version: 5.9.2
(...)
2017/12/19-18:55:01.089031 7f05c02ef700 [db/db_impl_write.cc:1220] [testCf] New memtable create with log file: #2821. Immutable memtables 1.
2017/12/19-18:55:01.089043 7f05c02ef700 [WARN] [db/column_family.cc:638] [testCf] Stopping writes because we have 2 immutable memtables (waiting for flush), max_write_buffer_number is set to 2
(...)
2017/12/19-18:55:01.127138 7f05bad5e700 [db/flush_job.cc:267] [testCf] [JOB 2813] Flushing memtable with next log file: 2821
(...)
2017/12/19-18:55:01.134796 7f05d0765700 [db/filesnapshot.cc:34] File Deletions Disabled
2017/12/19-18:55:01.146147 7f05d0765700 [WARN] [db/wal_manager.cc:426] [WalManager] /path/to/rocks/data/002821.log: dropping 32768 bytes; Corruption: checksum mismatch
2017/12/19-18:55:01.146160 7f05d0765700 [WARN] [db/wal_manager.cc:426] [WalManager] /path/to/rocks/data/002821.log: dropping 5962502 bytes; Corruption: error in middle of record
2017/12/19-18:55:01.146175 7f05d0765700 [WARN] [db/wal_manager.cc:426] [WalManager] /path/to/rocks/data/002821.log: dropping 32761 bytes; Corruption: missing start of fragmented record(1)
(...)
2017/12/19-18:55:01.149351 7f05d0765700 [db/filesnapshot.cc:57] File Deletions Enabled
(...)
2017/12/19-18:55:01.180045 7f05bad5e700 [db/db_impl_compaction_flush.cc:50] [JOB 2815] Syncing log #2821
(...)
2017/12/19-18:55:01.233180 7f05bad5e700 [db/db_impl_files.cc:403] [JOB 2815] Try to delete WAL files size 6034012, prev total WAL files size 12068024, number of live WAL files 2.
2017/12/19-18:55:01.233228 7f05bad5e700 [db/wal_manager:270] Move log file /path/to/rocks/data//002821.log to /path/to/rocks/data/archive/002821.log -- OK
```

**Comment by maysamyabandeh** [comment]
The call stack in your log must be GetSortedWalsOfType -> ReadFirstRecord -> ReadFirstLine but unfortunately I could not reproduce the problem on the master.

**Comment by ajkr** [comment]
Do you still see this problem?

**Comment by ktnowak** [comment]
No, we don't see this issue anymore. Seems it was resolved in version 5.13.1 - 'Fix WAL corruption caused by race condition between user write thread and backup/checkpoint thread.'

Thanks for help,

**Comment by solicomo** [comment]
FYI:
The first try of fix is #3603 .
And the second one is #3656 .
